crypto/cms/cms_smime.c

```diff
@@ -449,6 +412,108 @@ int CMS_verify(CMS_ContentInfo *cms, STACK_OF(X509) *certs,
             goto err;
 
     }
+
+    /*
+     * Handle CAdES signatures of higher levels (Baseline-T, -LT, and -LTA)
+     * that include additional information packed into the unsigned attributes
+     * of the signerInfo elements.
+     * Read this: they are not organized by document but by signer, therefore
+     * mulitple * objects may exist.
+     * Extract timestamp, if available, to set the correct time for certificate
+     * validation.
+     * Timestamp (and archive-timestamps) are taken from the unsigned
+     * attributes anyway, so it does not matter when they are evaluated.
+     * Evaluating earlier however supplies the timestamp and optional CRLs
+     * supporting the verification.
+     */
+    if (cadesVerify) {
+        int num, j;
+        for (i = 0; i < scount; i++) {
+            si = sk_CMS_SignerInfo_value(sinfos, i);
+
+            /*
+             * Look for embedded timestamp tokens and archiveTimestamps,
+             * note: there can be more than one
+             */
+            num = CMS_unsigned_get_attr_count(si);
+            if (num < 0)
+                continue;
+            for (j = 0; j < num; j++) {
+                X509_ATTRIBUTE *attr = CMS_unsigned_get_attr(si, j);
+                ASN1_OBJECT *obj = X509_ATTRIBUTE_get0_object(attr);
+                switch (OBJ_obj2nid(obj)) {
+                    case NID_id_smime_aa_timeStampToken:
+                        /*
+                         * Perform evaluation of timestamp. The value
+                         * timestamped is the original signature
+                         * If successfull, a valid signing time is available
+                         */
+                        if (!ossl_cms_handle_CAdES_SignatureTimestampToken(attr,
+                                 store, si->signature, &verification_time))
+                            goto err;
+                        vt = &verification_time;
+                        break;
+                    case NID_id_aa_ets_archiveTimestampV3:
+                        /*
+                         * Evaluate archiveTimestampV3 attribute
+                         */
+                        break;
```

Comment:
Should this code ignore `archiveTimestampV3` presence ?
IMHO, add a `ossl_assert()` or return an error.